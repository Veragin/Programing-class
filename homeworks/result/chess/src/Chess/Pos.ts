export const letters = ["A", "B", "C", "D", "E", "F", "G", "H"] as TY[];

type TPos = { x: TX; y: TY };
type TIndex = { i: number; j: number };

export class Pos {
    private _i: number;
    private _j: number;

    constructor(input: TPos | TIndex) {
        const inp = input as any;
        if (inp.x !== undefined) {
            this._i = inp.x - 1;
            this._j = letters.indexOf(inp.y);
        } else {
            this._i = inp.i;
            this._j = inp.j;
        }
    }

    get x() {
        return (this._i + 1) as TX;
    }
    get y() {
        return letters[this._j];
    }
    get i() {
        return this._i;
    }
    get j() {
        return this._j;
    }

    isEqual(pos: Pos) {
        return this.i === pos.i && this.j === pos.j;
    }

    export() {
        return [this._i, this._j];
    }
}
