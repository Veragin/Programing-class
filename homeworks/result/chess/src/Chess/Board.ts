import { BOARD_SIZE } from "./Const";
import { ChessPiece } from "./Pieces/ChessPiece";
import { Pos } from "./Pos";

export class Board {
    board: TBoard = [];

    constructor() {
        for (let i = 0; i < BOARD_SIZE; i++) {
            this.board.push([]);
            for (let j = 0; j < BOARD_SIZE; j++) {
                this.board[i].push(null);
            }
        }
    }

    set(piece: ChessPiece, pos: Pos) {
        try {
            const { i, j } = piece.findMyPosInBoard(this.board);
            this.board[i][j] = null;
        } catch {}

        this.board[pos.i][pos.j] = piece;
    }
}
