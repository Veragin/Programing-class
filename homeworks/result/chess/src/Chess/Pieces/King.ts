import { Pos } from "../Pos";
import { isOnBoard } from "../utils";
import { ChessPiece } from "./ChessPiece";

export class King extends ChessPiece {
    move(board: TBoard, myPos: Pos): Pos[] {
        const res: Pos[] = [];
        const i = myPos.i;
        const j = myPos.j;

        for (let oi = i - 1; oi < i + 2; oi++) {
            for (let oj = j - 1; oj < j + 2; oj++) {
                if ((oi === i && oj === j) || !isOnBoard(oi) || !isOnBoard(oj)) {
                    continue;
                }

                const p = board[oi][oj];
                if (p instanceof ChessPiece && p.player === this.player) {
                    continue;
                }

                res.push(new Pos({ i: oi, j: oj }));
            }
        }

        return res;
    }
}
