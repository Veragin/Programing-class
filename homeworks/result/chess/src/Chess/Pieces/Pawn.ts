import { Pos } from "../Pos";
import { ChessPiece } from "./ChessPiece";

export class Pawn extends ChessPiece {
    move(board: TBoard, myPos: Pos) {
        const res: Pos[] = [];
        const i = myPos.i;
        const j = myPos.j;

        const newI = this.player === "black" ? i - 1 : i + 1;

        if (board[newI][j] === null) {
            res.push(new Pos({ i: newI, j }));
        }

        let p = board[newI][j - 1];
        if (p instanceof ChessPiece && p.player !== this.player) {
            res.push(new Pos({ i: newI, j: j - 1 }));
        }

        p = board[newI][j + 1];
        if (p instanceof ChessPiece && p.player !== this.player) {
            res.push(new Pos({ i: newI, j: j + 1 }));
        }

        return res;
    }
}
