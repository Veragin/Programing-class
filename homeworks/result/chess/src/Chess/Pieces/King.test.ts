import { Board } from "../Board";
import { Pos } from "../Pos";
import { King } from "./King";
import { Pawn } from "./Pawn";

const createData = () => {
    const board = new Board();
    const king = new King("white");
    const pos = new Pos({ x: 2, y: "B" });

    board.board[pos.i][pos.j] = king;

    return {
        board: board.board,
        king,
        pos,
    };
};

describe("King", () => {
    it("can move", () => {
        const { board, king, pos } = createData();

        board[pos.i + 1][pos.j + 1] = new Pawn("black");

        const res = king.move(board, pos);
        expect(res.length).toBe(8);

        const exp = res.map((pos) => pos.export());
        expect(exp).toMatchSnapshot();
    });

    it("cannt move if is blocked by figure", () => {
        const { board, king, pos } = createData();
        board[pos.i][pos.j] = null;
        pos.i = 0;
        pos.j = 0;
        board[pos.i][pos.j] = king;

        const res1 = king.move(board, pos);
        expect(res1.length).toBe(3);
    });
});

export {};
