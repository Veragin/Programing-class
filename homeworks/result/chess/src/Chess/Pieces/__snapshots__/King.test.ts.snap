// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`King can move 1`] = `
Array [
  Array [
    0,
    0,
  ],
  Array [
    0,
    1,
  ],
  Array [
    0,
    2,
  ],
  Array [
    1,
    0,
  ],
  Array [
    1,
    2,
  ],
  Array [
    2,
    0,
  ],
  Array [
    2,
    1,
  ],
  Array [
    2,
    2,
  ],
]
`;
